### DISCUSSION TITLE
RGBImage or TwoPhotonSeries?

### DISCUSSION THREAD

**Discussion by yzhaoinuw** [dandi/helpdesk#92] [Q&A]
Hello folks, I have a question about choosing the right data class to store my image data in an NWB file. I got **z-stack** images taken by a two-photon microscope. Should I use _[RGBImage](https://nwb-schema.readthedocs.io/en/latest/format.html#rgbimage)_ or _[TwoPhotonSeries](https://nwb-schema.readthedocs.io/en/latest/format.html#sec-twophotonseries)_?

I currently used the _RGBImage_ class to store the images at different depth and then put all these _RGBImage_ in one [_Images_](https://nwb-schema.readthedocs.io/en/latest/format.html#images) class because the z-stack images are static images. For example, if I have z-stack images at 100 different depths, I would use 100 _RGBImage_ to store the image at each depth and then put them in one _Images_ class. And since my images have no more than three channels and the _RGBImage_ class requires three dimension, I used nan to fill the missing dimension if any.  I was able to write my data into nwb files this way and pass the validation of [NWBInspector](https://github.com/NeurodataWithoutBorders/nwbinspector).

But meanwhile, I also wonder if the _TwoPhotonSeries_ class is more appropriate. The problem is that it is a time series and requires time or frame in the first dimension. My data is not a time series but has depth. Can I use depth in place of frame? 

Which class would be the best practice for my data? Any advice or pointer is appreciated.

**Comment by CodyCBakerPhD** [comment]
I would suggest the `TwoPhotonSeries` despite the fact that it's static - you can always have a trivial length of `1` on the first axis (time) with the other axes being image width x image height x depth (z-stack)

A big reason for this suggestion is that the highly valuable related metadata about the optic channels, imaging plane structure, microscope device, etc. is required by the `TwoPhotonSeries` approach, but not by the RGB Image. It also allows much more precise annotation of the excitation wavelengths for each optic channel + indicator, whereas RGB is really intended to represent true RGB pixel colors in color (non-grayscale) image formats

**Reply by yzhaoinuw** [reply]
Hi @CodyCBakerPhD, thank you for the quick reply! 

My data has four dimension to begin with: (`nChannels`, `width`, `height`, `depth`). `nChannels` can be 1, 2, or 3, but no more than 3. Would you suggest that I create one `TwoPhotonSeries` for each channel (and the first dimension has a trivial legnth of 1 like you suggested) or create just one `TwoPhotonSeries` (and the first dimension is `nChannels`)?

The first option makes more sense to me, but I got the following critical issue warnings when running [NWBInspector](https://github.com/NeurodataWithoutBorders/nwbinspector) on the resulting nwb file:

```
0  CRITICAL
===========

0.0  z-stack-up.nwb: check_data_orientation - 'TwoPhotonSeries' object at location '/acquisition/TwoPhotonSeriesChan_3'
       Message: Data may be in the wrong orientation. Time should be in the first dimension, and is usually the longest dimension. Here, another dimension is longer.

0.1  z-stack-up.nwb: check_data_orientation - 'TwoPhotonSeries' object at location '/acquisition/TwoPhotonSeriesChan_2'
       Message: Data may be in the wrong orientation. Time should be in the first dimension, and is usually the longest dimension. Here, another dimension is longer.

0.2  z-stack-up.nwb: check_data_orientation - 'TwoPhotonSeries' object at location '/acquisition/TwoPhotonSeriesChan_1'
       Message: Data may be in the wrong orientation. Time should be in the first dimension, and is usually the longest dimension. Here, another dimension is longer.
``` 

I guess the warnings are due to the first dimension being 1, much less than the dimension of `width`, `height`, or `depth`. Are these warnings a big problem or can I ignore these?

**Reply by CodyCBakerPhD** [reply]
> Would you suggest that I create one TwoPhotonSeries for each channel (and the first dimension has a trivial legnth of 1 like you suggested)

Exactly, that is how NWB represents data from multiple color channels. Specifically, each series would point to a different ImagingPlane, and each ImagingPlane would have a different optic channel.


 > or create just one TwoPhotonSeries (and the first dimension is nChannels)?

The first dimension of every TimeSeries in NWB must always be time, according to [this Best Practice](https://nwbinspector.readthedocs.io/en/dev/best_practices/time_series.html#data-orientation), so I would not recommend doing it that way under any circumstance


> Are these warnings a big problem or can I ignore these?

Alas, that is the one `CRITICAL` rated warning from the Inspector that is implemented as a heuristic and so can be very wrong in cases such as this.

Yes, this message can be safely ignored in your case. With the CLI, you can do this with the extra flag `--ignore check_data_orientation`, or just pay it no mind in the report.

Also no worries on eventual problems with DANDI validatioj due to this - they downgrade that check to a `VIOLATION` so as not to prevent upload in cases like this. You can see the Inspector report using DANDI requirements with the the extra flag `--config DANDI`

**Reply by yzhaoinuw** [reply]
Hi @CodyCBakerPhD, thank you for your immediate reply again. You said 

> Specifically, each series would point to a different ImagingPlane, and each ImagingPlane would have a different optic channel. 

Could you explain why I would need to create an `ImagingPlane` for each `TwoPhotonSeries`, ie., channel? Based on my understanding, my data was all recorded on the same imaging plane. The channels literally mean different colors sensed by the microscope. I currently only created one `ImagingPlane`, and pointed data from each channel, in the shape of (1, `width`, `height`, `depth`), to this one `ImagingPlane`. Is this acceptable?

Regarding the `ImagingPlane`, do they correspond to the `depth` (z-stack) in my data? In other words, should I link the image at each depth
to a separate `ImagingPlane`? For example, if I arrange my data in such shape: (`depth`, `width`, `height`, `nChannels`), should I link `data(i, :, :, :)` to a separate `ImagingPlane` for `i=1:depth`?

**Reply by CodyCBakerPhD** [reply]
Sorry for the delay - I was out of town for a couple of days.

> Could you explain why I would need to create an ImagingPlane for each TwoPhotonSeries, ie., channel? Based on my understanding, my data was all recorded on the same imaging plane. The channels literally mean different colors sensed by the microscope. I currently only created one ImagingPlane, and pointed data from each channel, in the shape of (1, width, height, depth), to this one ImagingPlane. Is this acceptable?

It's a current limitation of the schema that an `OpticChannel` exist specific to each `ImagingPlane` (not as separate objects); so you would have one `ImagingPlane` for each color channel, all of the information about the physical plane (_e.g._, grid spacing) would be the same, but each would have a different `OpticChannel`. Then each `TwoPhotonSeries` would point to the `ImagingPlane` that has the correct `OpticChannel` for that series.


> Regarding the ImagingPlane, do they correspond to the depth (z-stack) in my data? In other words, should I link the image at each depth

This depends on the nature of the volume. As long as your depths overlap such that you can view a single `ImagingPlane` as a contiguous volume, then a single `ImagingPlane` per `OpticChannel` (but for all depths) is perfectly fine. I'd also give the plane the name 'Volume' somewhere in the title to indicate that, and/or include something about it in the description so it's really clear to someone looking at the file.

If they are non-overlapping however, then you would definitely want a separate `ImagingPlane` per depth.

Note there is wiggle room here and not a hard-and-fast rule; it's entirely based off of what you think is the better description of your experimental setup.
